Fix recursive madness

martong opened the pull request:
This PR attempts to solve all recursion related and "already imported" problems.

The original import logic should be really simple, we create the node, then we mark that as imported, then we recursively import the parts for that node and then set them on that node.
If we mark something as imported (`Imported()`) then we return with the corresponding `To` whenever we want to import that node again, this way recursive AST structures will be handled.
This PR tries to follow this logic.
To achieve this we must ensure that
- there are no `Import` calls in between any node creation (`::Create`) and the `Imported` call.
- there are no VisitXXX() calls in any other VisitYYY() function, only call of `Import()` allowed.
- we handle the already imported logic only in `Importer::Import(Decl*)`.

There are these degenarate cases when we may start an import both from the templated decl of a template and from the template itself.
In order to create a TemplateDecl we have to first import the templated decl, but during import of the template decl we have to import the template itself, bumm recursion.
We could limit the importer that it is possible to import only the template, but that would break the import of those nodes which directly refer anything from the templated decl.
So the solution is that we keep track the visited nodes from the `From` context.
Whenever we try to import a node which we already visited, we return the corresponding imported decl to that, or nullptr if there is no such.
This changes the interpretation of the return value of `Import(Decl*)` since we can return a nullptr during normal import as well. 
Perhaps using an optional<Decl*> would be more appropriate.

Fixes #282 , #274 , #323 , #312 , #311

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -384,7 +384,8 @@ AST_MATCHER_P(TemplateDecl, hasTemplateDecl,
 }
 
-TEST(ImportExpr, ImportParenListExpr) {
+// FIXME enable, handle SIGSEGV in clang::Decl::getKind(this=0x0)
```
Note, we receive here a SIGSEGV from the test code, the decl is successfully imported and can be dumped as well, still the matcher causes a segv.

Xazax-hun replied:
Is this introduced only after the changes? Maybe we set some attributes of some nodes to null?

martong replied:
Unfortunately this is introduced newly. Yes, perhaps some Decl is left as a nullptr. I'll try to debug with a sanitized build.

Xazax-hun replied:
I suspect this is a case where we already visited a decl but did not finish importing it and return a nullptr.

martong commented on unittests/AST/ASTImporterTest.cpp:
```diff
@@ -907,18 +908,63 @@ TEST(ImportDecl, DISABLED_ImportTemplateDefaultArgument) {
+}
+
+// Currently disabled, because lookup fails to find the existing templated decl.
+TEST_F(Fixture, DISABLED_ImportOfTemplatedDeclOfFunctionTemplateDecl) {
```
Note, this fails in the baseline too, there is a flaw in the current lookup logic, which is orthogonal to the changes of this PR.

balazske replied:
Probably this is what I try to solve right now. (Or the other direction: import of `FunctionDecl` that is a TemplatedDecl of a `FunctionTemplateDecl`.)

Xazax-hun commented on lib/AST/ASTImporter.cpp:
```diff
@@ -4206,36 +4204,31 @@ Decl *ASTNodeImporter::VisitClassTemplateDecl(ClassTemplateDecl *D) {
+      cast_or_null<CXXRecordDecl>(Importer.Import(FromTemplated));
+  if (!ToTemplated)
+    return nullptr;
+  Importer.Imported(FromTemplated, ToTemplated);
```
Wouldn't VisitRecordDecl mark this imported? Isn't this a redundant call?

martong replied:
Actually, this is redundant. But having redundant `Imported` calls is not a problem until the `To` is always the same.
Note w,e have an other redundancy in `Importer::Imprort(Decl*)` where we set the imported. In most cases this is redundant too, but better to have redundancy than missing just one case.

balazske replied:
Redundancy may not be bad, only here the call is immediately after the `Import` call and we (should) know that `Import` calls `Imported` always. (Otherwise there are a lot of places where `Import` is called and `Imported` is not called after it, somebody may think that here it is some special case.)

balazske replied:
Now I like it even less. Instead of removing a single `Imported` call it was inserted everywhere. One could ask: "Why is this all needed instead of doing it automatically in `Import`? What if the call to `Imported` is forgotten?" Alternative way of doing this: Do not call `Imported` but use an assertion to check if it was really imported.

martong replied:
Regarding to your first comment: We have to ensure that after a node is created with `::Create` we immediately add that to the `ImportedDecls`, or at least before we would recursively descend to other imports.

If we forget the call to `Imported` then we cannot simply track which decl is the pair of the already visited and possibly imported "from" decl in the "to" context.

balazske replied:
Yes the `Imported` needs to be called after the `Create` calls, and with no other `Import` in between. But if we import something with `Import` the `Imported` call is not required afterwards, I see it as redundant code. There can be special cases when `Imported` is called with an already existing node (multiple nodes from the From-context are mapped to a single node in the To-context).

Xazax-hun commented on lib/AST/ASTImporter.cpp:
```diff
@@ -6806,31 +6804,32 @@ Decl *ASTImporter::GetVAListTag(Decl *FromD) {
 }
 
 Decl *ASTImporter::Import(Decl *FromD) {
+
```
I prefer not to add new lines like this.

Xazax-hun commented on lib/AST/ASTImporter.cpp:
```diff
@@ -6806,31 +6804,32 @@ Decl *ASTImporter::GetVAListTag(Decl *FromD) {
   if (!FromD)
     return nullptr;
 
+  if (VisitedDecls.count(FromD) > 0)
```
Does the fix only work with VisitedDecls? If a Decl is visited it is not guaranteed that it is already finished with the import right? So we might return `nullptr` here in some cases which might cause failures. Do I miss something?

martong replied:
I think I answered all these questions in the summary of this PR.
Could you please read the summary of this PR?

martong replied:
The `VisitedDecls` is only needed to handle the templated decl vs described template circle.

Xazax-hun replied:
Oh, I see. Thanks. I am really uncomfortable about changing the meaning of the return value of Import call because the rest of the importer was built with a different assumption. I am afraid of this change causing lots of regressions.

balazske replied:
But the null return value indicated only error until now? We will have only more cases with non-null return value. (If null is returned in the new case and non-null in the old this should mean there is a recursive-import related problem.)

martong replied:
Okay, I think it may be the best to keep the original semantics of the return value: ie when nullptr there was an error.
I will add a new member to the importer with which we can use to query if we have encountered an already visited node during the import. This is ugly, but at least we could distinguish the error from the already visited state properly.

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -6806,31 +6804,32 @@ Decl *ASTImporter::GetVAListTag(Decl *FromD) {
     return nullptr;
 
+  if (VisitedDecls.count(FromD) > 0)
+    return GetAlreadyImportedOrNull(FromD);
```
Probably assert here instead of returning `nullptr` in not-found case?

martong replied:
Not good, in case of the templated vs described import cycle we may return with a nullptr.
E.g. during a FunctionTemplateDecl import we have to first import the templated decl, but during import of the template decl we have to import the template itself. The second time when we call the VisitFunctionTemplateDecl we return with a nullptr, since we already visited it but we don't have an imported definition to that (yet). 

Perhaps using an optional<Decl*> would be more appropriate, so we could distinguish error from the above situation. Or we could have a global pointer constant which has this meaning. @Xazax-hun what do you think?

balazske replied:
Probably assert in every case of error and return null pointer only if the node is under import but not yet created?

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -4206,36 +4204,31 @@ Decl *ASTNodeImporter::VisitClassTemplateDecl(ClassTemplateDecl *D) {
+      cast_or_null<CXXRecordDecl>(Importer.Import(FromTemplated));
+  if (!ToTemplated)
+    return nullptr;
+  Importer.Imported(FromTemplated, ToTemplated);
```
It's redundant, see my comment for Gabor's comment above.

balazske commented on lib/AST/ASTImporter.cpp:
```diff
@@ -7461,8 +7460,10 @@ void ASTImporter::CompleteDecl (Decl *D) {
 }
 
 Decl *ASTImporter::Imported(Decl *From, Decl *To) {
```
Is the return value used at all?

martong replied:
Of course, it is used all over in the importer, e.g when you import a body of a function then you set up the body on the node by using the returned value.

balazske replied:
I mean the return value of `Imported` not of `Import`. It is always the same as `To` and not really used.

martong replied:
ah okay, I'll try if we can make the return value `void`.

balazske commented:
Why not add a function "IsImportVisited" (or something similar) that returns the value of the previous "Visited" set (node was visited for import)? Or `IsImporting(Decl*)` that returns true if import of the node is currently in progress (`Import` was entered and not returned)? This is probably a less "tricky" solution.

martong commented:
@balazske I added `IsImporting(Decl*)` which now I use during the import of a templated decl. Also removed the `Imported()` from that point you guys pointed out.

martong commented:
Perhaps a better name would be `IsBeingImported()` instead of `IsImporting()`, at least based on this https://clang.llvm.org/doxygen/classclang_1_1TagType.html#ae8089b23a96d6656323f113a23c4e8cc

balazske commented:
`IsBeingImported` sounds better.
But the meaning should be clarified (add comment), is it true only for the time of the import (during the `Import` call) or remains it true afterwards (as it is now)? In the first case we should remove `FromD` from `VisitedDecls` before return from `Import` (or check `ImportedDecls` in `IsImporting`). In the second case `IsBeingImported` or `IsImporting` is not a good name. (Probably "WasVisitedDuringImport"?)

martong commented:
I realized the existing algorithm in `Import(Decl*)` does return null in some cases.
E.g. in `ImportDeclContext` the imported children can be null, that is why we have to check if that is not null:
```
    for (Decl *ToD : ListOfImportedDecls) { // O(n)
      if (ToD && ToD->getLexicalDeclContext() == ToDC) {
```
Also during importing types the imported decl can be also null. E.g.:
```
QualType ASTNodeImporter::VisitRecordType(const RecordType *T) {
  RecordDecl *ToDecl
    = dyn_cast_or_null<RecordDecl>(Importer.Import(T->getDecl()));
  if (!ToDecl)
    return QualType();

  return Importer.getToContext().getTagDeclType(ToDecl);
}
```
The point is, it looks like in the original algorithm as well, null is used to close the recursion and not just for error reporting.

martong commented:
I have simplified the patch, we don't need `isImported` at all, everything can be handled universally from `Import(Decl*)`.
Currently all except one unit test passes, but xerces does have regression so I mark this as WIP for now.

balazske commented on lib/AST/DeclCXX.cpp:
```diff
@@ -1521,8 +1521,9 @@ bool CXXRecordDecl::mayBeAbstract() const {
+    const RecordType* RecT = B.getType()->getAs<RecordType>();
+    assert(RecT);
+    CXXRecordDecl *BaseDecl = cast<CXXRecordDecl>(RecT->getDecl());
     if (BaseDecl->isAbstract())
```
This is only reformatting, not strict related to the other changes.
And `auto` can be used here as the type too.

martong commented:
I was working more on this and realized that indeed in some cases we can't just return null, because that would change dramatically the legacy behavior of the importer, e.g in case of the described-templated pendulum. 

So now I favor the existing pattern we already seen, i.e. to check before each ::Create whether we have created a Decl previously for that.

balazske commented:
The tests would be useful to add.
